cmd/snap/cmd_run.go

```diff
@@ -190,30 +210,66 @@ func maybeWaitForSecurityProfileRegeneration(cli *client.Client) error {
 	// is to fix the packaging so that snapd is stopped, upgraded
 	// and started.
 	//
-	// connect timeout for client is 5s on each try, so 12*5s = 60s
-	timeout := 12
-	if timeoutEnv := os.Getenv("SNAPD_DEBUG_SYSTEM_KEY_RETRY"); timeoutEnv != "" {
-		if i, err := strconv.Atoi(timeoutEnv); err == nil {
-			timeout = i
-		}
-	}
+	// connect timeout for client is 5s on each try, default retry count is 12,
+	// gives 60s of timeout
+	retryCount := getSystemKeyRetryCount()
 
 	logger.Debugf("system key mismatch detected, waiting for snapd to start responding...")
+	logger.Debugf("my system key: %v", my)
 
-	for i := 0; i < timeout; i++ {
-		// TODO: we could also check cli.Maintenance() here too in case snapd is
-		// down semi-permanently for a refresh, but what message do we show to
-		// the user or what do we do if we know snapd is down for maintenance?
-		if _, err := cli.SysInfo(); err == nil {
-			return nil
+	var rspError *client.Error
+
+	for i := 0; i < retryCount; i++ {
+		// refreshes the state of snapd maintenance internally
+		act, err := cli.SystemKeyMismatchAdvice(my)
+		if err == nil {
+			// we have an explicit response from snapd
+			switch act.SuggestedAction {
+			case client.MismatchActionProceed:
+				// we're good to go
+				logger.Debugf("continue execution")
+				return "", nil
+			case client.MismatchActionWaitForChange:
+				// snapd sent us an ID of a change to wait for
+				logger.Debugf("snapd started a regenerate profiles change with ID: %v", act.ChangeID)
+				return act.ChangeID, nil
+			default:
+				return "", fmt.Errorf("internal error: unexpected advice: %q", act.SuggestedAction)
+			}
+		}
+
+		// an error, which could be antything from socket being down, to
+		// relevant method, or our system-key version not being supported by the
+		// API if we're talking to an older snapd, or snapd is simply updating
+		restarting := false
+		if maintErr, ok := cli.Maintenance().(*client.Error); ok && maintErr.Kind == client.ErrorKindDaemonRestart {
+			restarting = true
+		}
+
+		if !restarting && errors.Is(err, client.ErrMismatchAdviceUnsupported) {
+			// snapd does not support system-key mismatch resolution, but it is
+			// responding so profiles must have been regenerated during startup
```

Comment:
I'm not sure about the 2nd sentence here, as it still the case that mounts could have happened since that restart?